docs/randomized_testing.md

```diff
@@ -168,6 +168,250 @@ roughly the same things from the user (which operations to run, how to generate
+they don't need to take state as an argument.
+
+```go
+func TestWorkload(t *testing.T) {
+    // Create the simulator.
+    s, err := sim.New(sim.Options{...})
+    if err != nil {
+        t.Fatal(err)
+    }
+
+    // Run the simulator for 10 seconds.
+    results, err := s.Run(10*time.Second, func(s *sim.Sim) error {
+        // State is a local variable.
+        replies := map[int]int{}
+
+        // Register a fake value.
+        sim.RegisterFake[B](s, &fakeb{})
+
+        // Register an op to call A.A.
+        sim.RegisterOp(s, sim.Op[int]{
+            Name: "CallA",
+            Gen:  func(*rand.Rand) int { ... },
+            Func: func(ctx context.Context, x int, a A) error {
+                if y, err := a.A(ctx, x); err == nil {
+                    replies[x] = y
+                }
+                return nil
+            },
+        })
+
+        // Register an op to call B.B.
+        sim.RegisterOp(s, sim.Op[struct{}]{
+            Name: "CallB",
+            Gen:  func(*rand.Rand) struct{} { ... },
+            Func: func(ctx context.Context, _ struct{}, b B) error {
+                b.B(ctx)
+                return nil
+            },
+        })
+
+        // Run a single simulation.
+        return s.Run()
+    })
+    if err != nil {
+        t.Fatal(err)
+    }
+}
+```
+
+**Pros.**
+The main benefit of this approach is that state is represented as local
+variables, and fakes are registered by value.
+
+**Cons.**
+This API has two distinct simulator types: one to run a bunch of simulations and
+one to run a single simulation. The API cannot prevent someone from registering
+different state, fakes, and ops for every individual simulation, which is not
+intended. More generally, it's a bit odd that you keep re-registering everything
+for every individual simulation.
+
+### Option 3
+
+Option 3 is shown below. This option wraps all the logic of the workload into
+the `Workload` struct, which is passed to the `sim.New[T]` function. The
+`Workload` struct has `weaver.Ref`s to the components needed in the test. It
+also has fields for any state.
+
+- Method `FakeB` returns a fake for component `B`. Generally, method `FakeFoo`
+  returns a fake for component `Foo`.
+- Method `GenCallA` generates the random values passed to `CallA`. Generally,
+  `GenFoo` generates the random values passed to `Foo`.
+- Methods `CallA` and `CallB` are the body of the ops. Generally, all exported
+  methods that don't begin with `Fake` or `Gen` are ops.
+
+```go
+type Workload struct {
+    a weaver.Ref[A]
+    b weaver.Ref[B]
+    replies map[int]int
+}
+
+func (w *Workload) Init(context.Context) error {
+    w.replies = map[int]int{}
+    return nil
+}
+
+func (w *Workload) FakeB(context.Context) B {
+    return &fakeb{}
+}
+
+func (w *Workload) CallA(ctx context.Context, x int) error {
+    if y, err := w.a.Get().A(ctx, x); err == nil {
+        w.replies[x] = y
+    }
+    return nil
+}
+
+func (*Workload) GenCallA(r *rand.Rand) int {
```

Comment:
We discussed possibly defining a generator as a different type, like we did for routing.

Reply:
Replied to your other comment with an example of this proposal.